diff --git a/configure.ac b/configure.ac
index 2fe24ce1..37914361 100644
--- a/configure.ac
+++ b/configure.ac
@@ -753,7 +753,7 @@ then
     freerdp_version="(2.x)"
     have_freerdp=yes
     PKG_CHECK_MODULES([RDP], [freerdp2 freerdp-client2 winpr2],
-                      [CPPFLAGS="${RDP_CFLAGS} -Werror $CPPFLAGS"]
+                      [CPPFLAGS="${RDP_CFLAGS} -Werror -Wno-error=deprecated-declarations $CPPFLAGS"]
                       [AS_IF([test "x${FREERDP_PLUGIN_DIR}" = "x"],
                              [FREERDP_PLUGIN_DIR="`$PKG_CONFIG --variable=libdir freerdp2`/freerdp2"])],
                       [freerdp_version=
@@ -765,7 +765,7 @@ then
     freerdp_version="(3.x)"
     have_freerdp=yes
     PKG_CHECK_MODULES([RDP], [freerdp3 freerdp-client3 winpr3],
-                      [CPPFLAGS="${RDP_CFLAGS} -Werror $CPPFLAGS"]
+                      [CPPFLAGS="${RDP_CFLAGS} -Werror -Wno-error=deprecated-declarations $CPPFLAGS"]
                       [AS_IF([test "x${FREERDP_PLUGIN_DIR}" = "x"],
                              [FREERDP_PLUGIN_DIR="`$PKG_CONFIG --variable=libdir freerdp3`/freerdp3"])],
                       [AC_MSG_WARN([
diff --git a/src/protocols/rdp/Makefile.am b/src/protocols/rdp/Makefile.am
index dc9638e3..7d993ddf 100644
--- a/src/protocols/rdp/Makefile.am
+++ b/src/protocols/rdp/Makefile.am
@@ -131,7 +131,7 @@ noinst_HEADERS =                                 \
     user.h
 
 libguac_client_rdp_la_CFLAGS = \
-    -Werror -Wall -Iinclude    \
+    -Werror -Wall -Wno-error=deprecated-declarations -Iinclude    \
     @COMMON_INCLUDE@           \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@          \
@@ -165,7 +165,7 @@ libguac_common_svc_client_la_SOURCES =        \
     plugins/guac-common-svc/guac-common-svc.c
 
 libguac_common_svc_client_la_CFLAGS = \
-    -Werror -Wall -Iinclude           \
+    -Werror -Wall -Wno-error=deprecated-declarations -Iinclude           \
     @LIBGUAC_INCLUDE@                 \
     @RDP_CFLAGS@
 
@@ -187,7 +187,7 @@ libguacai_client_la_SOURCES =           \
     plugins/ptr-string.c
 
 libguacai_client_la_CFLAGS = \
-    -Werror -Wall -Iinclude  \
+    -Werror -Wall -Wno-error=deprecated-declarations -Iinclude  \
     @COMMON_INCLUDE@         \
     @COMMON_SSH_INCLUDE@     \
     @LIBGUAC_INCLUDE@        \
diff --git a/src/protocols/rdp/channels/disp.c b/src/protocols/rdp/channels/disp.c
index da1ca800..0293843c 100644
--- a/src/protocols/rdp/channels/disp.c
+++ b/src/protocols/rdp/channels/disp.c
@@ -85,6 +85,10 @@ static void guac_rdp_disp_channel_connected(rdpContext* context,
     if (strcmp(args->name, DISP_DVC_CHANNEL_NAME) != 0)
         return;
 
+    /* Abort if display update module is not yet initialized */
+    if (guac_disp == NULL)
+        return;
+
     /* Init module with current display size */
     guac_rdp_disp_set_size(guac_disp, rdp_client->settings,
             context->instance, guac_rdp_get_width(context->instance),
@@ -127,6 +131,10 @@ static void guac_rdp_disp_channel_disconnected(rdpContext* context,
     if (strcmp(args->name, DISP_DVC_CHANNEL_NAME) != 0)
         return;
 
+    /* Abort if display update module is not yet initialized */
+    if (guac_disp == NULL)
+        return;
+
     /* Channel is no longer connected */
     guac_disp->disp = NULL;
 
diff --git a/src/protocols/rdp/input-queue.c b/src/protocols/rdp/input-queue.c
index 76994a8e..2088fd88 100644
--- a/src/protocols/rdp/input-queue.c
+++ b/src/protocols/rdp/input-queue.c
@@ -78,8 +78,7 @@ static void guac_rdp_handle_mouse_event(guac_rdp_client* rdp_client,
     /* If button mask unchanged, just send move event */
     if (mask == rdp_client->mouse_button_mask) {
         pthread_mutex_lock(&(rdp_client->message_lock));
-        GUAC_RDP_CONTEXT(rdp_inst)->input->MouseEvent(
-                GUAC_RDP_CONTEXT(rdp_inst)->input, PTR_FLAGS_MOVE, x, y);
+        freerdp_input_send_mouse_event(GUAC_RDP_CONTEXT(rdp_inst)->input, PTR_FLAGS_MOVE, x, y);
         pthread_mutex_unlock(&(rdp_client->message_lock));
     }
 
@@ -102,8 +101,7 @@ static void guac_rdp_handle_mouse_event(guac_rdp_client* rdp_client,
             if (released_mask & 0x04) flags |= PTR_FLAGS_BUTTON2;
 
             pthread_mutex_lock(&(rdp_client->message_lock));
-            GUAC_RDP_CONTEXT(rdp_inst)->input->MouseEvent(
-                    GUAC_RDP_CONTEXT(rdp_inst)->input, flags, x, y);
+            freerdp_input_send_mouse_event(GUAC_RDP_CONTEXT(rdp_inst)->input, flags, x, y);
             pthread_mutex_unlock(&(rdp_client->message_lock));
 
         }
@@ -121,8 +119,7 @@ static void guac_rdp_handle_mouse_event(guac_rdp_client* rdp_client,
 
             /* Send event */
             pthread_mutex_lock(&(rdp_client->message_lock));
-            GUAC_RDP_CONTEXT(rdp_inst)->input->MouseEvent(
-                    GUAC_RDP_CONTEXT(rdp_inst)->input, flags, x, y);
+            freerdp_input_send_mouse_event(GUAC_RDP_CONTEXT(rdp_inst)->input, flags, x, y);
             pthread_mutex_unlock(&(rdp_client->message_lock));
 
         }
@@ -133,16 +130,16 @@ static void guac_rdp_handle_mouse_event(guac_rdp_client* rdp_client,
             /* Down */
             if (pressed_mask & 0x08) {
                 pthread_mutex_lock(&(rdp_client->message_lock));
-                GUAC_RDP_CONTEXT(rdp_inst)->input->MouseEvent(
-                        GUAC_RDP_CONTEXT(rdp_inst)->input, PTR_FLAGS_WHEEL | 0x78, x, y);
+                freerdp_input_send_mouse_event(GUAC_RDP_CONTEXT(rdp_inst)->input,
+                        PTR_FLAGS_WHEEL | 0x78, x, y);
                 pthread_mutex_unlock(&(rdp_client->message_lock));
             }
 
             /* Up */
             if (pressed_mask & 0x10) {
                 pthread_mutex_lock(&(rdp_client->message_lock));
-                GUAC_RDP_CONTEXT(rdp_inst)->input->MouseEvent(
-                        GUAC_RDP_CONTEXT(rdp_inst)->input, PTR_FLAGS_WHEEL | PTR_FLAGS_WHEEL_NEGATIVE | 0x88, x, y);
+                freerdp_input_send_mouse_event(GUAC_RDP_CONTEXT(rdp_inst)->input,
+                        PTR_FLAGS_WHEEL | PTR_FLAGS_WHEEL_NEGATIVE | 0x88, x, y);
                 pthread_mutex_unlock(&(rdp_client->message_lock));
             }
 
diff --git a/src/protocols/rdp/keyboard.c b/src/protocols/rdp/keyboard.c
index b6058989..75d84975 100644
--- a/src/protocols/rdp/keyboard.c
+++ b/src/protocols/rdp/keyboard.c
@@ -109,8 +109,8 @@ static void guac_rdp_send_key_event(guac_rdp_client* rdp_client,
 
     /* Send actual key */
     pthread_mutex_lock(&(rdp_client->message_lock));
-    GUAC_RDP_CONTEXT(rdp_inst)->input->KeyboardEvent(
-            GUAC_RDP_CONTEXT(rdp_inst)->input, flags | pressed_flags, scancode);
+    freerdp_input_send_keyboard_event(GUAC_RDP_CONTEXT(rdp_inst)->input,
+            flags | pressed_flags, scancode);
     pthread_mutex_unlock(&(rdp_client->message_lock));
 
 }
@@ -139,8 +139,8 @@ static void guac_rdp_send_unicode_event(guac_rdp_client* rdp_client,
 
     /* Send Unicode event */
     pthread_mutex_lock(&(rdp_client->message_lock));
-    GUAC_RDP_CONTEXT(rdp_inst)->input->UnicodeKeyboardEvent(
-            GUAC_RDP_CONTEXT(rdp_inst)->input, 0, codepoint);
+    freerdp_input_send_unicode_keyboard_event(GUAC_RDP_CONTEXT(rdp_inst)->input,
+            0, codepoint);
     pthread_mutex_unlock(&(rdp_client->message_lock));
 
 }
@@ -169,8 +169,8 @@ static void guac_rdp_send_synchronize_event(guac_rdp_client* rdp_client,
 
     /* Synchronize lock key states */
     pthread_mutex_lock(&(rdp_client->message_lock));
-    GUAC_RDP_CONTEXT(rdp_inst)->input->SynchronizeEvent(
-            GUAC_RDP_CONTEXT(rdp_inst)->input, flags);
+    freerdp_input_send_synchronize_event(GUAC_RDP_CONTEXT(rdp_inst)->input,
+            flags);
     pthread_mutex_unlock(&(rdp_client->message_lock));
 
 }
diff --git a/src/protocols/rdp/tests/Makefile.am b/src/protocols/rdp/tests/Makefile.am
index e1e22e36..214059fd 100644
--- a/src/protocols/rdp/tests/Makefile.am
+++ b/src/protocols/rdp/tests/Makefile.am
@@ -38,7 +38,7 @@ test_rdp_SOURCES =      \
     fs/normalize_path.c
 
 test_rdp_CFLAGS =                \
-    -Werror -Wall -pedantic      \
+    -Werror -Wall -pedantic -Wno-error=deprecated-declarations \
     @LIBGUAC_CLIENT_RDP_INCLUDE@ \
     @LIBGUAC_INCLUDE@
 
